Dependency versioning

jacg opened the pull request:
An attempt at addressing #495.

The idea is:

+ Include a dependency configuration tag in a committed file (`manage.sh` for now).
+ Povide `manage.sh` commands which
   + create conda environments which have the tag in their name and contain the correct set of dependencies
   + switch conda environment
+ Make the `manage.sh` testing commands automatically switch to the conda env which is appropriate for the currently checked out code.
+ The environments are made lazily: no environment is created until it is needed, which will only happen if
   + A commit which needs that environment is checked out, and
   + The user asks for the environment to be created, ether
      + explicitly, or
      + implicitly, by using `manage.sh` to run tests.

The tags are, for now, of the form `YYYY-MM-DD`, but we could use relase tag names instead. It might be a good idea to update the dependencies on every release: that way we'll never get too far behind the march of progress.

An example: If you clone and install the repo containing this branch today, you will create a conda env called `IC-3.6-2018-06-30`. If you then check out a commit which corresponds to an earlier dependency configuration, as soon as you run the tests with `source manage.sh run_tests(_par)`, a second conda environment called `IC-3.6-2018-05-03` will be created and activate implicitly. Checking out the tip of the branch once more, and running the tests (with `manage.sh`) will automatically switch to the first environment.

The whole lot needs to be cleaned up, made more robust, and the database should be updated automatically too. We also need to think about how to deal with commits which predate the addition of this feature.

I can imagine that better approaches could be invented. Let me know what you think.

It seems that we don't have the manpower to think about this carefully and the only way to prevent this from being ingnored and festering forever, is to approve and commit the first thing that looks like it works.

jacg committed:
Update pinned pcakage versions to most recent

jacg committed:
source activate -> conda activate

jacg committed:
Bump conda version

jacg committed:
Adapt paolina to evolution of NetworkX interface from v1 to v2

`all_pairs_dijkstra_path_length` has become lazy: it used to return a
dictionary. Now it returns an iterator.

Graph query methods have become lazy too, so `nodes_iter` becomes the
default `iter`

jacg committed:
Rephrase find_extrema_and_length

jacg committed:
Heed FutureWarnings about DataFrame.groupby()

The meaning of tuples passed to groupby is going to change in the
future: they will no longer specify multiple keys, but will be
understood to be a single key themselves. To specify multiple keys, a
list should be used instead.

jacg committed:
Turn work_in_python_version into a one-stop solution

Setting up the environment before working on or using IC was a bit of
a PITA because you were required to use different `manage.sh`
commands, depending on whether

+ conda is already installed
+ the cond environment you want to use already exists

The `manage.sh` function `work_in_python_version` has been mae more
intelligent: it now discovers which of these steps need to be taken,
and should be usable as a single entry point to getting going with IC.

This commit also contains the beginning of dependency versioning:
the IC conda environments will now have names which include a version
stamp (rather than just the Python version being used). This should
allow us to switch to an environment which contains the dependency
versions which were in place at the time any given version of your
code was committed.

jacg committed:
Run tests in parallel by default work_in_python_version

jacg committed:
Add work_in_python_version_no_test

By default we want the tests to run, to highlight pre-existing
problems when someone starts to work in/on IC. However, as this work
is leading up to automatic environment switching when historical
commits are checked out, we do not want to run all the tests on each
checkout. So we need a version that switches environment without
running the tests.

jacg committed:
Ensure that PYTHON_VERSION is set sooner in manage.sh

jacg committed:
Make manage.sh testing automatically choose appropriate env

jacg committed:
Prevent run_tests_par from exiting shell

jacg committed:
Deduce PYTHON_VERSION from env if not passed as arg

The manage.sh testing functions do not accept a python version as an
argument. But they need to know this version, now that they need to
switch automatically to the environment which corresponds to the
checked out code.

jacg committed:
Update quickstart guide

jjgomezcadenas committed:
Fix signal.decimate

The change from Scipy 0.2 to Scipy 1.1 caused a nasty error due to a
parameter (n) which changed its default value. This is fixed by
setting explicitely n to 30 (which was the default value in 0.2).

jjgomezcadenas committed:
Add two tests to document the decimate error.

One test shows the problem and the other one its solution.

jacg commented on invisible_cities/sierpe/fee_test.py:
```diff
@@ -53,6 +53,73 @@ def test_fee_params():
+    The downsampling factor.
+
+    n : int, optional
+    The order of the filter (1 less than the length for ‘fir’).
```
The signal to noise ratio here is awful. Rather than copypasting the whole manual entries for the two versions, I would just describe the change: `n` used to have these defaults, now it has these, and maybe leave a link to the manual.

jacg commented on invisible_cities/sierpe/fee_test.py:
```diff
@@ -53,6 +53,73 @@ def test_fee_params():
+
+    try:
+        assert 23 < adc_to_pes     < 23.1
+    except AssertionError:
```
`except Whatever: pass` is not helpful at all. It fails to convey the meaning and is full of holes.

While I totally agree with the intent and motivation of this test and its partner, I don't think that there is any sensible way of implementing it, because to demonstrate the change requires running tests in two different conda envs containing two different versions of scipy, in a single execution of the test suite, which just isn't going to happen.

I propose that we just bump the version of scipy and fix the concomitant bug in a single commit with a good commit message, and forget about the test that demonstrates. The in-step evolution of the test and version number will just have to do as documentation.

jjgomezcadenas replied:
I can change pass by a print that states that the error is skipped on purpose. The test demonstrates the problem independently of the conda environment, because in the first conda environment it works and in the second fails but is caught. I do believe that documenting the problem is useful. Actually I included the test because I was sure you were going to request it.

jacg replied:
I was going to reply, but now I see that you have removed it. OK, let's leave it like that. We have more urgent things to do.

jjgomezcadenas committed:
Remove test_show_scipy_nasty_feature.

Too verbose and difficult to understand. Instead,
test_show_signal_decimate_signature should be enough to document the issue.

jjgomezcadenas committed:
replace which conda by conda --version

Command which conda does not work in MacOS. Using conda --version
works and returns the current version of conda, which may be
nice. Without this, the script attempts to install conda every time in MacOS.

jacg commented on manage.sh:
```diff
@@ -89,6 +89,9 @@ dependencies:
 - notebook=5.5.0
+- jupyterlab=0.32.1=py36_0
+- jupyterlab_launcher=0.10.5=py36_0
+- python.app=2=py36_8
```
Tests fail because it doesn't like `python.app==2=py36_8`

jacg committed:
Stop failing to run parallel tests in work_in_python_version

The `manage.sh` command line argument sometimes specifies the python
version, sometimes the number of processors to use for parallel
testing. `work_in_python_version` accepts a python version, but
internally it calls run_tests_par, which requires the command line
argument to specify the number of processors. This conflict resulted
in the parallel tests not being run at all.

This is a quick hack to fix the problem as quickly as possible.

jjgomezcadenas reviewed (approved):
I have tried the system and works. It may create a bit of hassle when conflicting with the old version of conda, but it can be easily solved.
